### ISSUE TITLE
Sharing backbone model for different task adapters

### ISSUE THREAD

**Issue by jedzill4** [huggingface/peft#854]
I've successfully trained two LoRA adapters for BERT models: one for token classification and another for text classification. Currently, I'm able to load each adapter individually using `AutoPertModelForTokenClassification` and `AutoPertModelForSequenceClassification` respectively. However, I'm wondering if there's a way to efficiently load the BERT model for both of these tasks simultaneously.

The reason for my inquiry is that I need to deploy these models on machines with very limited resources, so sharing the backbone between the two models (and adapters) would be great. Has anyone faced a similar scenario?

**Comment by BenjaminBossan** [comment]
AFAIK, there is no functionality in PEFT to do what you asked. IIUC, the whole model is identical for both tasks, only the classification head is different. It would probably be possible with a bit of custom code to write a function that replaces the classification head with either of the ones you need, while keeping the backbone identical. Switching adapters should already work and aforementioned function would have to ensure that the correct adapter is enabled when there is a switch.

Alternatively, but this would require a bit more engineering, you could detach the head completely and only serve the outputs from the backbone in one process, and then send it to one of two separate processes, the first serving the token classification head, the other the sequence classification head.

However, what would be extremely difficult with either solution is switching adapters (and heads) while also batching at the same time. So if you have limited resources, since you wouldn't be able to use batching, the proposed solutions might not be practical for you.

**Comment by julien-gadonneix** [comment]
Any update on this?

**Comment by BenjaminBossan** [comment]
There is nothing in PEFT that would automatically do this and it is honestly outside of scope for PEFT. However, if I had a similar problem, I would choose a different approach today, namely loading the base model weights only once and for the second base model, load the empty weights on `meta` device using accelerate's `init_empty_weights`, then share the weights between the two models. Below I show a POC:

```python
import os

import torch
from transformers import AutoModelForCausalLM, AutoModelForSequenceClassification
from peft import get_peft_model, LoraConfig, TaskType, PeftModel
from accelerate.big_modeling import init_empty_weights

path = "/tmp/peft/shared-weights"
device = "cuda:0"
model_id = "facebook/opt-125m"
model_id = "meta-llama/Llama-3.2-1B"
model = None

# create the causal LM if needed
if not os.path.exists(path + "/clm/adapter_model.safetensors"):
    print("Creating CLM model...")
    model = AutoModelForCausalLM.from_pretrained(model_id, device_map=device)
    config = LoraConfig(task_type=TaskType.CAUSAL_LM)
    model = get_peft_model(model, config)
    model.save_pretrained(path + "/clm")

# create the sequence classification model if needed
if not os.path.exists(path + "/cls/adapter_model.safetensors"):
    print("Creating CLS model...")
    model = AutoModelForSequenceClassification.from_pretrained(model_id, device_map=device)
    config = LoraConfig(task_type=TaskType.SEQ_CLS)
    model = get_peft_model(model, config)
    model.save_pretrained(path + "/cls")

del model
torch.cuda.empty_cache()

print("Loading CLM model")
model_clm = AutoModelForCausalLM.from_pretrained(model_id, device_map=device)
print(f"Max CUDA memory allocated after loading CLM model: {torch.cuda.max_memory_allocated(device) / 2** 20:.2f} MB")

print("Loading empty CLS model on meta device")
with init_empty_weights():
    model_cls = AutoModelForSequenceClassification.from_pretrained(model_id)
# if we replace this with:
# model_cls = AutoModelForSequenceClassification.from_pretrained(model_id, device_map=device)
# the max memory allocated will be ~962 MB
print(f"Max CUDA memory allocated after loading empty CLS model: {torch.cuda.max_memory_allocated(device) / 2** 20:.2f} MB")

print("sharing weights")
# there is probably a more elegant way to achieve this
for name, param in model_clm.named_parameters():
    # exclude lm_head, which is not part of the CLS model
    # this is technically not needed here because lm_head is skipped since it shares its weight with the embedding
    if name != "lm_head.weight":
        parent_name, _, attr_name = name.rpartition(".")
        parent = model_cls.get_submodule(parent_name)
        setattr(parent, attr_name, param)

# populate the score layer of the classification model with data
model_cls.score.weight = torch.nn.Parameter(torch.randn(model_cls.score.weight.shape, device=device))
print(f"Max CUDA memory allocated after populating CLS score layer: {torch.cuda.max_memory_allocated(device) / 2** 20:.2f} MB")

print("Loading PEFT adapters")

model_clm = PeftModel.from_pretrained(model_clm, path + "/clm")
model_cls = PeftModel.from_pretrained(model_cls, path + "/cls")
print(f"Max CUDA memory allocated after loading PEFT adapters: {torch.cuda.max_memory_allocated(device) / 2** 20:.2f} MB")

x = torch.randint(10, 100, (1, 50)).to(device)
print("Running CLM model")
with torch.no_grad():
    out_clm = model_clm(x, output_hidden_states=True)
    assert torch.isfinite(out_clm.logits).all(), "CLM model output is not float"

print("Running CLS model")
with torch.no_grad():
    out_cls = model_cls(x, output_hidden_states=True)
    assert torch.isfinite(out_cls.logits).all(), "CLS model output is not float"

print(f"Max CUDA memory allocated after running inference: {torch.cuda.max_memory_allocated(device) / 2** 20:.2f} MB")

print("test finished successfully")
```

For me, this prints:

```
Loading CLM model
Max CUDA memory allocated after loading CLM model: 4719.25 MB
Loading empty CLS model on meta device
Some weights of LlamaForSequenceClassification were not initialized from the model checkpoint at meta-llama/Llama-3.2-1B and are newly initialized: ['score.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Max CUDA memory allocated after loading empty CLS model: 4719.25 MB
sharing weights
Max CUDA memory allocated after populating CLS score layer: 4719.25 MB
Loading PEFT adapters                                                                                                                                                                                                         
Max CUDA memory allocated after loading PEFT adapters: 4725.06 MB
Running CLM model
Running CLS model
Max CUDA memory allocated after running inference: 4806.95 MB
test finished successfully
```

(note: run the script twice, as the first time around, it needs to create the PEFT adapters)

When I don't share the weights, I get this instead:

```
Loading CLM model
Max CUDA memory allocated after loading CLM model: 4719.25 MB
Loading empty CLS model on meta device
Some weights of LlamaForSequenceClassification were not initialized from the model checkpoint at meta-llama/Llama-3.2-1B and are newly initialized: ['score.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Max CUDA memory allocated after loading empty CLS model: 9435.51 MB
Max CUDA memory allocated after populating CLS score layer: 9435.51 MB
Loading PEFT adapters                                                                                                                                                                                                   
Max CUDA memory allocated after loading PEFT adapters: 9441.31 MB
Running CLM model
Running CLS model
Max CUDA memory allocated after running inference: 9523.21 MB
test finished successfully
```

So sharing the weights seems to work. Be aware that this is just a POC, when you apply the idea to your own code, double-check that the expected results are returned.

**Comment by julien-gadonneix** [comment]
Thanks ! I'll try your solution and let you know.
Here is another option, do you think it is correct?

```
n = 10
backbone = Model.from_pretrained("")
config = LoraConfig(
    r=r,
    lora_alpha=lora_alpha,
    target_modules=["query", "value"],
    lora_dropout=lora_dropout,
    bias="none",
)
self.base = get_peft_model(base, config)
self.base.add_adapter("branch0", config)
for i in range(1, n):
    self.base.add_adapter(f"branch{i}", config)
```

And then:

```
x = torch.random.randn((32, 10, 128))
for i in range(n):
    self.base.set_adapter(f"branch{i}")
    features = self.base(x)
```

**Comment by BenjaminBossan** [comment]
@julien-gadonneix What you show is using the same base model with multiple different LoRA adapters. This will be memory efficient because the base model is only loaded once. Note, however, that this is not what was asked originally: OP asked about using slightly different base models (token classification vs sequence classification) that share the same backbone. That is more complicated, which is why I posted the snippet above. If the base model is the same, as seems to be the case in your example, this complexity is indeed not required.
